### ISSUE TITLE
CA1806 DoNotIgnoreMethodResults should generate warning for IDisposable return values

### ISSUE THREAD

**Issue by qmfrederik** [dotnet/roslyn-analyzers#1057]
#### Analyzer package

`Microsoft.Maintainability.Analyzers`
#### Analyzer

CA1806: `DoNotIgnoreMethodResults`
#### Repro steps

Consider the following line:

```
cancellationToken.Register(() => socket.Dispose());
```

where [`CancellationToken.Register`](https://msdn.microsoft.com/nl-be/library/dd321635%28v=vs.110%29.aspx) returns a `CancellationTokenRegistration`, which implements `IDisposable`.
#### Expected behavior

CA1806 does not generate a warning, apparently it only triggers warnings for very specific methods.

However, I believe that CA1806 could benefit from generating a warning if:
- The method returns a value
- That value implements `IDisposable`

This seems to be a clear indication that the return value should be disposed of by the caller; hence the caller should use the method result.
#### Actual behavior

CA1806 does not generate a warning.

**Comment by mavasani** [comment]
This will be a compat break, but seems a reasonable thing to complain about.

**Comment by Evangelink** [comment]
@mavasani I have started to have a look at this extension for the rule and I am wondering whether:

1/ I should exclude `Task` from the disposable cases or only some of the assert methods.

2/ Also handle `IAsyncDisposable`

**Comment by Evangelink** [comment]
Ping @mavasani

**Comment by Evangelink** [comment]
ping @mavasani

**Comment by mavasani** [comment]
> This seems to be a clear indication that the return value should be disposed of by the caller; hence the caller should use the method result.

This is by design and implementing this suggestion would lead to a lot of false positives.

1. We have [CA2000](https://docs.microsoft.com/visualstudio/code-quality/ca2000) to cover flagging dispose leaks. Implementing this suggestion would likely lead to overlap of lot of cases where both CA2000 and CA1806 get flagged, which is undesirable.
2. CA2000 is not flagged for the case in issue description as it is very common to have helper methods that return a disposable object without dispose ownership transfer. The dispose ownership could still be with the callee and the caller does not need to invoke dispose on the object. CA2000 has been tuned to only flag dispose leaks when we know for sure that the method being invoked is a _dispose creation_ where there is a dispose ownership transfer. Flagging all methods that return an IDisposable as a dispose leak would be a false positive that leads to double dispose.

**Comment by Evangelink** [comment]
Your comment makes sense! 

This rule seems like a good candidate for the rule "aggressiveness" threshold I once mentioned to you. When dealing with some C++ resources I'd like to be able to have some option to change the threshold between false-positives and false-negatives so that even if I end-up reviewing more cases I am more confident I haven't missed cases where the value should have been disposed. But that would clearly be a different task.
